### ISSUE TITLE
CUDA: Provide memset (or memset-like) functionality as part of the public interface

### ISSUE THREAD

**Issue by gmarkall** [numba/numba#5595]
`numba.cuda.cudadrv.driver` provides the [`device_memset`](https://github.com/numba/numba/blob/master/numba/cuda/cudadrv/driver.py#L2266) function that calls either `cuMemsetD8Async` or `cuMemsetD8` depending on whether a stream is supplied.

It might be nice to provide a more public interface for setting memory. One could expose a fairly raw interface, like:

```python
def memset(dst, val, itemsize, count, stream=0):
   """Memset on the device. Parameters:
    - dst: A device array or other CUDA memory object
    - val: Value to assign
    - itemsize: size in bytes of the assigned item - e.g. 4 for int32
    - count: Number of items to set
    - stream: An optional stream - a nonzero stream implies an asynchronous memset"""
```

Although this matches expectations of a `memset` interface coming from a C background, it may be safer to provide the same functionality by slice assignment, e.g.:

```python
x = cuda.device_array(1024, dtype=np.int32)
x[:] = 7
```

where the number of items and data type are automatically determined, and memsetting of a portion of a device array is achieved by slicing the array and then assigning to the whole slice. At present the syntax `x[:] = 7` does work, but is implemented in a fairly inefficient manner - when running with `NUMBA_CUDA_LOG_LEVEL=DEBUG`, it results in:

```
== CUDA [17788] DEBUG -- call driver api: cuCtxGetCurrent
== CUDA [17788] DEBUG -- call driver api: cuCtxGetDevice
== CUDA [17789] DEBUG -- call driver api: cuMemAlloc_v2
== CUDA [17790] DEBUG -- call driver api: cuMemcpyHtoD_v2
== CUDA [18035] DEBUG -- call driver api: cuLinkCreate_v2
== CUDA [18035] DEBUG -- call driver api: cuLinkAddData_v2
== CUDA [18035] DEBUG -- call driver api: cuLinkComplete
== CUDA [18035] DEBUG -- call driver api: cuModuleLoadDataEx
== CUDA [18036] DEBUG -- call driver api: cuModuleGetFunction
== CUDA [18036] DEBUG -- call driver api: cuFuncGetAttribute
== CUDA [18036] DEBUG -- call driver api: cuFuncGetAttribute
== CUDA [18036] DEBUG -- call driver api: cuFuncGetAttribute
== CUDA [18036] DEBUG -- call driver api: cuFuncGetAttribute
== CUDA [18036] DEBUG -- call driver api: cuFuncGetAttribute
== CUDA [18036] DEBUG -- call driver api: cuLinkDestroy
== CUDA [18036] DEBUG -- call driver api: cuOccupancyMaxPotentialBlockSize
== CUDA [18036] DEBUG -- call driver api: cuLaunchKernel
== CUDA [18037]  INFO -- add pending dealloc: cuMemFree_v2 8 bytes
```

because Numba creates and jits a kernel to assign the RHS, as well as copying the assigned scalar to the device, when it could have been done more efficiently with a call to `cuMemsetD32` for this particular example. So, implementing the second option comes down to optimising the implementation of [`DeviceNDArray._do_setitem()`](https://github.com/numba/numba/blob/master/numba/cuda/cudadrv/devicearray.py#L577) to use an appropriate `cudaMemset*` function wherre possible.

**Comment by sklam** [comment]
IIRC, we stopped at making use of memset because the handling for strided ndim array gets challenging quickly.

**Comment by gmarkall** [comment]
Indeed, it's hard to use memset for all cases - but those where the slice we're assigning to is C-contiguous should be safe to use memset for, should they not?

It seems like this would make a large improvement for a common case, replacing 15 driver API calls with one (though of course one needs benchmarks to make sure, but it shouldn't be too difficult to set up a reasonable measurement).

**Comment by sklam** [comment]
Yes, any contiguous array would be safe and very likely faster with memset.

**Comment by gabri94** [comment]
I am currently using a custom kernel written by me, to initialize an ndim array with a specific value. It is quite simple, but think that a direct call to memset could be more efficient.
Specifically i have a kernel which do some calculation in-memory and at the end of the execution i need to erase the memory and set it back to a custom value.

Is there any clean way to do it on an n-dimensional array? I am not that deep into numba, so I don't know if all ndim array are strided or even whether they are C-contiguous or F-contiguous by default

**Comment by seibert** [comment]
ndim arrays on the GPU can have any of the layouts you describe, so the fast CUDA memset could be used for C and Fortran order, and the slower kernel-based memset used for strided arrays.
